### PR TITLE
Allow registering new backends with decorators

### PR DESCRIPTION
**Pull request by hellock**

### CODE REVIEW

File: mmcv/fileio/file_client.py [hunk 1]
```diff
@@ -222,16 +222,35 @@ def __init__(self, backend='disk', **kwargs):
...
+        if not isinstance(name, str):
+            raise TypeError('the backend name should be a string, '
+                            f'but got {type(name)}')
         if not inspect.isclass(backend):
             raise TypeError(
                 f'backend should be a class but got {type(backend)}')
         if not issubclass(backend, BaseStorageBackend):
             raise TypeError(
                 f'backend {backend} is not a subclass of BaseStorageBackend')
+        if not force and name in cls._backends:
```
Comment (Johnson-Wang) [inline mmcv/fileio/file_client.py:235]: assert hasattr(cls, '_backends')

File: mmcv/fileio/file_client.py [hunk 2]
```diff
@@ -222,16 +222,35 @@ def __init__(self, backend='disk', **kwargs):
         self.client = self._backends[backend](**kwargs)
 
     @classmethod
-    def register_backend(cls, name, backend):
+    def _register_backend(cls, name, backend, force=False):
```
Comment (Johnson-Wang) [inline mmcv/fileio/file_client.py:225]: docstring

File: mmcv/fileio/file_client.py [hunk 3]
```diff
@@ -222,16 +222,35 @@ def __init__(self, backend='disk', **kwargs):
...
 
         cls._backends[name] = backend
 
+    @classmethod
+    def register_backend(cls, name, backend=None, force=False):
+        if backend is not None:
+            cls._register_backend(name, backend, force=force)
+            return
+
+        def _register(in_cls):
```
Comment (Johnson-Wang) [inline mmcv/fileio/file_client.py:248]: `in_cls` --> `in_backend` for clarity.

File: tests/test_fileclient.py [hunk 4]
```diff
@@ -203,6 +218,57 @@ def get_text(self, filepath):
...
+
+            def get_text(self, filepath):
+                return 'text3'
+
+        example_backend = FileClient('example3')
+        assert example_backend.get(self.img_path) == 'bytes3'
+        assert example_backend.get_text(self.text_path) == 'text3'
+        assert 'example3' in FileClient._backends
+
+        with pytest.raises(KeyError):
```
Comment (Johnson-Wang) [inline tests/test_fileclient.py:252]: Comment on the specific error

File: mmcv/fileio/file_client.py (hunk 1 above)
Comment (hellock) [reply mmcv/fileio/file_client.py:235]: This is always true, for both FileClient and subclasses.

File: mmcv/fileio/file_client.py [hunk 5]
```diff
@@ -222,16 +222,72 @@ def __init__(self, backend='disk', **kwargs):
...
+
+                def get(self, filepath):
+                    return filepath
+
+                def get_text(self, filepath):
+                    return filepath
+
+        Args:
+            name (str): The name of the registered backend.
+            backend (class, optional): The backend class to be registered.
```
Comment (Johnson-Wang) [inline mmcv/fileio/file_client.py:275]: :class:`BaseStorageBackend` would be better?

**Pull request merged**